l10n_id_taxform_period/models/tax_period.py

```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Prime Force Indonesia
+# Copyright 2016 OpenSynergy Indonesia
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models, fields, api
+from openerp.tools.translate import _
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+
+
+class TaxYear(models.Model):
+    _name = "l10n_id.tax_year"
+    _description = "Tax Year"
+    _order = "date_start asc, id"
+
+    name = fields.Char(
+        string="Tax Year",
+        required=True,
+    )
+    code = fields.Char(
+        string="Code",
+        required=True,
+    )
+    date_start = fields.Date(
+        string="Date Start",
+        required=True,
+    )
+    date_end = fields.Date(
+        string="Date End",
+        required=True,
+    )
+    period_ids = fields.One2many(
+        string="Periods",
+        comodel_name="l10n_id.tax_period",
+        inverse_name="year_id",
+    )
+
+    @api.constrains("date_start", "date_end")
+    def _check_range(self):
+        if self.date_end <= self.date_start:
+            strWarning = _("The start date must precede it's end date")
+            raise models.ValidationError(strWarning)
```

Comment:
I tested it with cProfile, here's the result

`
>>> def test2():
...     raise exception.ValidationError("test2")
... 
>>> cProfile.run('test2()')
         3 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <stdin>:1(test2)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
`
`
>>> def test():
...     raise models.ValidationError("error test")
... 
>>> cProfile.run('test()')
         5 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <stdin>:1(test)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 exceptions.py:33(__init__)
        1    0.000    0.000    0.000    0.000 exceptions.py:68(__init__)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
`